auacm/angular/src/app/competition.service.ts

```diff
@@ -0,0 +1,341 @@
+   * a user stops viewing a scoreboard for a competition.
+   */
+  resetScoreboard() {
+    this.competition = new Competition();
+    this.competitionSource.next(this.competition);
+  };
+
+  /**
+   * Fetches the competition details for a competition.
+   * 
+   * @param cid - the competition id
+   */
+  fetchCompetition(cid: number) {
+    this.getCompetition(cid).then(competition => {
+      this.competition = competition;
+      this.competitionSource.next(this.competition);
+    });
+  };
+
+  /**
+   * Returns the client time offset for a competition.
+   * 
+   * @returns the client-side time offset of a competition.
+   */
+  getClientTimeOffset(): number {
+    return this.clientTimeOffset;
+  };
+
+  /**
+   * Creates a new competition.
+   * 
+   * @param competition - the competition to create
+   * @param problems - the problems for the competition
+   * 
+   * @returns the new competition
+   */
+  createCompetition(competition: Competition, problems: Problem[]) : Promise<Competition> {
+    var problemData = [];
+    for (var i = 0; i < problems.length; i++) {
+      problemData.push({
+        label: String.fromCharCode("A".charCodeAt(0) + i),
+        pid: problems[i].pid
+      });
+    }
+    var formData = new URLSearchParams();
+    formData.append('name', competition.name);
+    formData.append('start_time', competition.startTime.toString());
+    formData.append('length', competition.length.toString());
+    formData.append('problems', JSON.stringify(problemData));
+    // in python, a non empty string that is parsed will be true,
+    // while an empty string will be false.
+    formData.append('closed', competition.closed ? '1' : '');
+
+    var headers = new Headers();
+    headers.append('Content-Type', 'application/x-www-form-urlencoded');
+
+    return new Promise((resolve, reject) => {
+      this._http.post('/api/competitions', formData.toString(), { headers: headers }).subscribe((res: Response) => {
+        if (res.status === 200) {
+          resolve(new Competition());
+        } else {
+          resolve(undefined);
+        }
+      }, (err: Response) => {
+        resolve(undefined);
+      })
+    });
+  };
+
+  /**
+   * Returns a map of comptitions and their data.
+   * 
+   * @returns a map of competitions
+   */
+  getAllCompetitions() : Promise<Map<string, Competition[]>> {
+    return new Promise((resolve, reject) => {
+      this._http.get('/api/competitions').subscribe((res: Response) => {
+        var competitions = new Map<string, Competition[]>();
+        var data = res.json().data;
+        if (res.status == 200) {
+          for (var competitionType in data) {
+            competitions[competitionType] = [];
+            for (var i = 0; i < data[competitionType].length; i++) {
+              // I'm unsure what is returned, going to check for all parameters     - John
+              var competition = new Competition();
```

Comment:
If you want, you can make a list of field names, and iterate over that:

`var fields = ['cid', 'closed', ....];
 for (var j = 0; j < fields.length; j++) {
     if(data[competitionType][i][fields[j]] != undefined) competition[fields[j]] = data[competitionType][i][fields[j]];
 }
`